File: concepts/data structures/implementation/tree/bt_traversal/depth_first/python/bt_pre_order.py
from bt_base.python.binary_tree import BinaryNode


def walk(curr, path):
    if curr is None:
        return path

    # Recursion
    # Pre
    path.append(curr.value)

    # Recurse
    walk(curr.left, path)
    walk(curr.right, path)

    # Post
    return path


def pre_order_search(head):
    return walk(head, [])


def test_pre_order_search():
    tree = BinaryNode(
        1,
        BinaryNode(2, BinaryNode(4), BinaryNode(5)),
        BinaryNode(3),
    )
    result = pre_order_search(tree)

    print(f"Pre-order: {result} | Expected: [1, 2, 4, 5, 3]")


test_pre_order_search()

File: concepts/data structures/implementation/tree/bt_traversal/depth_first/typescript/bt_pre_order.ts
import { BinaryNode } from "../../../bt_base/typescript/binary_tree";

function walk(curr: BinaryNode<number> | null, path: number[]): number[] { 
  if (!curr) return path;

  // recursion
  // pre
  path.push(curr.value);
  
  // recurse
  walk(curr.left, path);
  walk(curr.right, path);

  // post
  return path;
}

function pre_order_search(head: BinaryNode<number>): number[] {
  return walk(head, []);
}

function testPreOrderSearch(): void {
  const tree = new BinaryNode(
    1,
    new BinaryNode(2, new BinaryNode(4), new BinaryNode(5)),
    new BinaryNode(3),
  );
  const result = pre_order_search(tree);

  console.log(`Pre-order: ${result.join(", ")} | Expected: 1, 2, 4, 5, 3`);
}

testPreOrderSearch();
